### PR TITLE
{?exists} and {^exists} now supports values from a promise

### PR DESCRIPTION
**Pull request by samuelms1**
I logged this as bug #752 - {?exists/} behavior different when using promises.

### CODE REVIEW

**Comment by sethkinast** [comment]
I think this won't quite be right because when you use `await` in this way, you'll be shifting the context of the body. This is very close, though.

I think something like this might fail with your change:

```
{
  a: {
    b: {
      test: Promise.resolve("BAD")
    },
    test: Promise.resolve("GOOD")
  }
}

{#a}{?b}{test}{/b}{/a}```

For the test runner failure, it's because Node 0.10 doesn't have Promise-- use the "FalsePromise" constructor at the top of the test file.

**Comment by sethkinast** [comment]
You probably want to use `chunk.map` similar to how `Chunk#await` does it, but you need to only check the truthiness of the body rather than delegating to `Chunk#section`, which is what `await` will do.

It hits this code:

```
if (bodies) {
  chunk = chunk.section(data, context, bodies);
} 
```

**Comment by samuelms1** [comment]
@sethkinast thanks for the comments!  I got pretty busy yesterday, so I had to switch over to other work.  I'm going to take another look.

**Comment by sethkinast** [comment]
Yep no rush! I appreciate you taking the time to work on a patch :)

**Comment by samuelms1** [comment]
@sethkinast can you take another look?  You are right that using `Chunk#await` resulted in the context changing to the value of the resolved promise which is not desirable.  I think I have that addressed now, but let me know what you think.

File: lib/dust.js [hunk 1]
```diff
@@ -973,25 +995,34 @@
...
-        }
+        dust.log('Unhandled promise rejection in `' + context.getTemplateName() + '`', INFO);
+        chunk.renderError(err, context, bodies).end();
       });
     });
   };
 
   /**
+   * Render an error body if available
+   * @param thenable {Thenable} the target thenable to await
```
Comment (sethkinast) [inline lib/dust.js]: wrong param

File: lib/dust.js [hunk 2]
```diff
@@ -973,25 +995,34 @@
...
+   * @param context {Context} context to use to render the error
+   * @param bodies {Object} may optionally contain an "error" which will be rendered
+   * @return {Chunk}
+   */
+  Chunk.prototype.renderError = function(err, context, bodies) {
```
Comment (sethkinast) [inline lib/dust.js]: Nice, good reusable func

File: lib/dust.js [hunk 3]
```diff
@@ -861,6 +861,17 @@
     var body = bodies.block,
         skip = bodies['else'];
 
+    if (dust.isThenable(elem)) {
```
Comment (sethkinast) [inline lib/dust.js]: Could you turn this into a little reusable func that we can use for exists and notexists both? Look above at `getWithResolvedData` as an example-- just a little helper func not attached to the Chunk prototype.

Comment (sethkinast) [inline lib/dust.js]: by "this" I meant the code inside the if() block that is

File: lib/dust.js [hunk 4]
```diff
@@ -725,6 +725,11 @@
     return this;
   };
```
Comment (samuelms1) [inline lib/dust.js]: I added these docs straight from http://www.dustjs.com/docs/helper-api/.  Looks like other docs are missing from some chunk methods that are on the helper-api page.  I could add these if they aren't considered bloat.  This Chunk#map method is great and the docs really made it's usage clear to me.  As I was looking at making the code smaller / more resuable, it seemed apparent that what I really needed was a new map method for thenables, so I created the mapThenable(...) method.  I'd like to put it on the Chunk object, but that would be altering the public API of that and I'm not sure that desirable.  I'm also wondering if I should move the Chunk#renderError method off of Chunk for the same reason or if it's simple, generic, and useful enough to keep.

File: lib/dust.js (hunk 3 above)
Comment (samuelms1) [inline lib/dust.js]: The code was definitely redundant between the two methods.  I made a `mapThenable` method to address it.

File: lib/dust.js (hunk 2 above)
Comment (samuelms1) [inline lib/dust.js]: Thanks.  Should I move it off of the Chunk object?  When I added it originally I wasn't considering that I was updating a public api that's pretty well documented here: http://www.dustjs.com/docs/helper-api/

File: lib/dust.js [hunk 5]
```diff
@@ -861,6 +895,12 @@
     var body = bodies.block,
         skip = bodies['else'];
 
+    if (dust.isThenable(elem)) {
+      return mapThenable(this, elem, context, bodies, function(chunk, data){
```
Comment (sethkinast) [inline lib/dust.js]: Fantastic. Style nit: space between `function(chunk, data)` and its brace

**Comment by sethkinast** [comment]
Great work! If I can hassle you to fix the brace nit (which I should add to the linter...) we're ready to roll.

Also shoutout for writing great function documentation.

File: lib/dust.js (hunk 4 above)
Comment (sethkinast) [inline lib/dust.js]: I think you have the right separation. I think `mapThenable` is not something that a general public API would necessarily need, but `renderError` is nice. We can add it to the docs.

File: lib/dust.js [hunk 6]
```diff
@@ -1025,11 +1074,8 @@
           if(ended) {
             return;
           }
-          if(errorBody) {
-            chunk.render(errorBody, context.push(err));
-          } else {
-            dust.log('Unhandled stream error in `' + context.getTemplateName() + '`', INFO);
-          }
+          chunk.renderError(err, context, bodies);
+          dust.log('Unhandled stream error in `' + context.getTemplateName() + '`', INFO);
```
Comment (samuelms1) [inline lib/dust.js]: Note that this is a logic difference.  Previously `dust.log(...)` was only called when there was no render body whereas now it is always logged.  I think it's ok.

**Comment by sethkinast** [comment]
Thanks again for the PR! 🎆

**Comment by samuelms1** [comment]
@sethkinast it was my pleasure.  It was a lot of fun.  Makes me want to contribute more in open source projects.

What is the release cycle for these changes?  I'd like to start using them as soon as they become available.

**Comment by sethkinast** [comment]
I'm tempted to cut a 2.8, which I didn't plan on before a 3, and spend 3.0 working on an ES6 rewrite.

You could use the changes today by putting a commit hash in your `package.json` instead of a version number, if you wanted.

**Comment by samuelms1** [comment]
@sethkinast We don't reference dustjs-linkedin directly or I might consider that.  We use hoffman for the expressjs integration it provides.  hoffman uses duster and duster uses dustjs-linkedin.

The ES6 rewrite sounds like a fun project to work on.  Is there a branch for it?

**Comment by sethkinast** [comment]
I have not pushed a public branch for ES6 yet but hopefully soon.

Let me see about cutting a 2.8 in the next week or two.

**Pull request merged**